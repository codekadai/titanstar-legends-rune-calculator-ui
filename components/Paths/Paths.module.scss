.paths {
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  gap: 1.5rem;
}

.emptyPaths {
  margin: 0;
  text-align: center;
  font-size: 1rem;
  font-weight: 500;
}

@media (min-width: 768px) {
  .emptyPaths {
    text-align: left;
  }
}
