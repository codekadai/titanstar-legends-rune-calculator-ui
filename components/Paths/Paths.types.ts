import { RuneType } from "../Rune/Rune.types";

export type PathType = RuneType[];
