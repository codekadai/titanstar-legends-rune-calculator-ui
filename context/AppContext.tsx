"use client";

import { getPaths, getPlayers } from "@/api";
import { PathType } from "@/components/Path/Path.types";
import React, {
  createContext,
  FC,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";

export interface AppContextProviderProps {
  children: React.ReactNode;
}

interface AppContextType {
  currentSpentPoints: number;
  currentPaths: PathType[];
  talentPoints: number;
  isLoading: boolean;
  setCurrentSpentPoints: React.Dispatch<React.SetStateAction<number>>;
  setCurrentPaths: React.Dispatch<React.SetStateAction<PathType[]>>;
  setTalentPoints: React.Dispatch<React.SetStateAction<number>>;
  setIsLoading: React.Dispatch<React.SetStateAction<boolean>>;
}

const defaultContextValue: AppContextType = {
  currentSpentPoints: 0,
  currentPaths: [],
  talentPoints: 0,
  isLoading: false,
  /* v8 ignore next */
  setCurrentSpentPoints: () => {},
  /* v8 ignore next */
  setCurrentPaths: () => {},
  /* v8 ignore next */
  setTalentPoints: () => {},
  /* v8 ignore next */
  setIsLoading: () => {},
};

const AppContext = createContext<AppContextType>(defaultContextValue);

export const AppProvider: FC<AppContextProviderProps> = ({ children }) => {
  const [currentSpentPoints, setCurrentSpentPoints] = useState<number>(0);
  const [currentPaths, setCurrentPaths] = useState<PathType[]>([]);
  const [talentPoints, setTalentPoints] = useState<number>(0);
  const [isLoading, setIsLoading] = useState<boolean>(true);

  useEffect(() => {
    const fetchPaths = async () => {
      try {
        const paths = await getPaths();
        setCurrentPaths(paths);
      } catch (error) {
        console.error("getPaths at AppContext: ", error);
        return [];
      } finally {
        setIsLoading(false);
      }
    };
    const fetchPlayer = async () => {
      try {
        const players = await getPlayers();
        setTalentPoints(players[0].talentPoints);
      } catch (error) {
        console.error("getPlayers at AppContext: ", error);
        return [];
      } finally {
        setIsLoading(false);
      }
    };
    fetchPaths();
    fetchPlayer();
  }, []);

  const contextValue = useMemo(
    () => ({
      currentSpentPoints,
      currentPaths,
      talentPoints,
      isLoading,
      setCurrentSpentPoints,
      setCurrentPaths,
      setTalentPoints,
      setIsLoading,
    }),
    [currentSpentPoints, currentPaths, talentPoints, isLoading]
  );

  return (
    <AppContext.Provider value={contextValue}>{children}</AppContext.Provider>
  );
};

export const useAppContext = () => useContext(AppContext);
