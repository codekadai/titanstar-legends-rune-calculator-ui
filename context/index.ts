export * from "./AppContext";
export * from "./Providers";
