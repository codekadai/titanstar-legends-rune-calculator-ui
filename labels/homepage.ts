export const homepage = {
  title: "TitanStar Legends - Rune Mastery Loadout Talent Calculator 9000",
};
