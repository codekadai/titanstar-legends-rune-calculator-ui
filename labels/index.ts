export * from "./metadata";
export * from "./homepage";
export * from "./tree";
export * from "./points";
export * from "./paths";
export * from "./loader";
