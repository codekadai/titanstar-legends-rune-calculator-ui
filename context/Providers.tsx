"use client";

import { AppProvider } from "./AppContext";

export const Providers = ({ children }: { children: React.ReactNode }) => {
  return <AppProvider>{children}</AppProvider>;
};
