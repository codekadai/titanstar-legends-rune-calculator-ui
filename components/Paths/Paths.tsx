import { useAppContext } from "@/context";
import styles from "./Paths.module.scss";
import { Path } from "../Path";
import type { PathType } from "./Paths.types";
import { paths } from "@/labels";

export const Paths = () => {
  const { currentPaths } = useAppContext();

  return (
    <div className={styles.paths}>
      {currentPaths.length !== 0 ? (
        currentPaths.map((path: PathType, pathIndex: number) => (
          <Path key={pathIndex} path={path} pathIndex={pathIndex} />
        ))
      ) : (
        <p className={styles.emptyPaths}>{paths.noPaths}</p>
      )}
    </div>
  );
};
