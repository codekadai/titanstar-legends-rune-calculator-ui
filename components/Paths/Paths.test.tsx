import React from "react";
import { describe, it, expect, vi, Mock } from "vitest";
import { render, screen } from "@testing-library/react";
import { Paths } from "./Paths";
import { useAppContext } from "../../context";
import { paths } from "../../labels";

vi.mock("@/context", () => ({
  useAppContext: vi.fn(),
}));

describe("Paths tests", () => {
  it("shows a message when there are no paths", () => {
    (useAppContext as Mock).mockReturnValue({ currentPaths: [] });

    render(<Paths />);

    const noPathsMessage = screen.getByText(paths.noPaths);
    expect(noPathsMessage).toMatchSnapshot();
  });
});
