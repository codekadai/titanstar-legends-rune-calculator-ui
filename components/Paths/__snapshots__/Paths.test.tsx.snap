// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Paths tests > shows a message when there are no paths 1`] = `
<p
  class="_emptyPaths_cb9680"
>
  Sorry, there's no talent path available at this moment
</p>
`;
